File: pa2/Lex.c
// Spenser Estrada spmestra@ucsc.edu
// pa2
//
//-----------------------------------------------------------------------------
// Lex.c
// Illustrates file input-output commands and text processing using the
// string functions strtok and strcat.
//-----------------------------------------------------------------------------

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include "List.h"

#define MAX_LEN 160

int main(int argc, char * argv[]){

   int count=0;//, n;
   FILE *in, *in2, *out;
   char line[MAX_LEN];

   // check command line for correct number of arguments
   if( argc != 3 ){
      printf("Usage: %s <input file> <output file>\n", argv[0]);
      exit(1);
   }

   // open files for reading and writing 
   in = fopen(argv[1], "r");
   out = fopen(argv[2], "w");
   if( in==NULL ){
      printf("Unable to open file %s for reading\n", argv[1]);
      exit(1);
   }
   if( out==NULL ){
      printf("Unable to open file %s for writing\n", argv[2]);
      exit(1);
   }

   // counts total number of lines in input file
   while( fgets(line, MAX_LEN, in) != NULL)  {
      count++;
   }
   printf("Count: %d\n", count);
   fclose(in);
   
   // copies individual lines of file to array
   in2 = fopen(argv[1], "r");
   if( in2==NULL ){
      printf("Unable to open file %s for reading\n", argv[1]);
      exit(1);
   }
   char words[count][MAX_LEN];
   int i = 0;
   while( fgets(line, MAX_LEN, in2) != NULL ){
      strncpy(words[i], line, strlen(line)+1);
      i++;
   }
   
   List L = newList();
   for(i = 0; i<count; i++){
      append(L, i);
   }
   
   //sorts indices of words[][] into lexicograpic order
   int j;
   for(j = 1; j < length(L); j++){
      moveTo(L, j);
      int key = getElement(L);
      
      movePrev(L);
      while (getIndex(L) >= 0 && strcmp(words[getElement(L)], words[key])>0 ) {
         movePrev(L);
      }
      if (getIndex(L) < 0) prepend(L, key);
      else insertAfter(L, key);
      moveTo(L, j+1);
      delete(L);
   }
   
   // close files
   fclose(in2);

   // prints element of words[][] corresponding to sorted list
   moveTo(L, 0);
   while( getIndex(L) >= 0 ){
      //printf("%s", words[getElement(L)]);
      fprintf(out, "%s", words[getElement(L)]);
      moveNext(L);
   } 
   
   fclose(out);
   
   // empty list and free Nodes/List
   clear(L);
   freeList(&L);
   
   return(0);
}

File: pa1/Lex.java
// Spenser Estrada
// spmestra@unix.ucsc.edu
// pa1

import java.io.*;
import java.util.Scanner;

class Lex {
   public static void main(String[] args) throws IOException{
      Scanner scan = null;
      PrintWriter print = null;
      int lineNumber = 0;

      if(args.length < 2){
         System.out.println("Usage: FileIO infile outfile");
         System.exit(1);
      }

      scan = new Scanner(new File(args[0]));
      print = new PrintWriter(new FileWriter(args[1]));

      while( scan.hasNextLine() ) {
         lineNumber++;
         scan.nextLine();
      }
      scan.close();

      Scanner scan2 = new Scanner(new File(args[0]));

      String[] words = new String[lineNumber];
      for (int i = 0; i < words.length; i++) {
         words[i] = scan2.nextLine();
      } 
     
      List sort = new List();
      for (int i = 0; i < words.length; i ++)
         sort.append(i);
      
      
      inSort(words, sort);
      
      for (sort.moveTo(0); sort.getIndex() >= 0; sort.moveNext() ) {
         print.write( words[ sort.getElement() ]+"\n" );
      }
      
      scan2.close();
      print.close();
   }

   static void inSort(String[] A, List L) {
      for (int j = 1; j < L.length(); j++) {
         L.moveTo(j);
         int key = L.getElement();
         L.movePrev();
         while (L.getIndex() >= 0 && (A[L.getElement()].compareTo( A[key] ) >0 ) ) {
            L.movePrev();
         }
         if (L.getIndex() < 0) L.prepend(key); 
         else L.insertAfter(key);
         L.moveTo(key+1);
         L.delete();
      }
   }
}
